src/librbd/ImageCtx.h

```diff
@@ -301,8 +301,13 @@ namespace librbd {
                         bool thread_safe);
 
     ExclusiveLock<ImageCtx> *create_exclusive_lock();
+    void destroy_exclusive_lock(ExclusiveLock<ImageCtx> *exclusive_lock, bool closed=true);
```

Comment:
I'd prefer to just utilize the existing ```put()``` methods on the objects where necessary

Reply:
The reference count for the object is 2 (as `open()/init()` would be been done). We would essentially need to invoke `put()` twice to delete the object (e.g https://github.com/ceph/ceph/blob/master/src/librbd/image/CloseRequest.cc#L114 does not invoke `close()` to decrement the reference count). I'm thinking if we can still retain our own custom reference counting class for that...

Reply:
Not sure I understand the provided link. Why would the reference count be 2 and why wouldn't the reference count be decremented as needed?

Reply:
We start with a reference count a 1 (in constructor) and increment it again after `open()`. In the example provided above, `m_image_ctx->object_map` is not closed and deleted directly.

Reply:
What ```open``` method? ```ObjectMap<I>::open```? I think the reference counting should be consistently applied. The state machines should only ```get``` / ```put``` when they have an async operation in-flight.  Therefore, in the case of ```ObjectMap```, the ref count would be 1 just after creation, would be 2 while ```ObjectMap::init``` is in-flight, and would be back down to 1 after the ```init``` context callback is invoked.

In the linked example, I think those ```delete``` statements need to be yanked / replaced w/ a ```put```.

Reply:
I had thought about incrementing the reference count on each async operation, but then it seemed incrementing in `init()` (or whatever the initialization routine is) would guarantee that the object doesn't get deleted till the equivalent `shut_down()` method is invoked (as the operations would be serialized by `ImageState` class).

Reply:
The end goal here is to avoid the need to have a special "helper" method to properly release the reference count. The logically flow of the object should reflect the correct reference count.

Reply:
ok -- this is what I'm following now (as you had mentioned in the original tracker issue):

Reference count in `create_context_callback` -- Callers would need to pass `create_context_callback()` the object to be reference counted. This would mean a handful number of changes all around librbd (but the reference counting is done at exactly one central place). Although, reference counting in the public API methods would probably have less number of changes, reference counting in a central place would be clean.